src/raft.rs

```diff
@@ -2225,4 +2535,364 @@ impl<T: Storage> Raft<T> {
         m.request_snapshot = self.pending_request_snapshot;
         self.send(m);
     }
+
+    // Pick the delegate for sending raft logs to a single member in a group
+    // Only used in `send_append`
+    //
+    // If no delegate is found, send raft logs directly
+    // If the target needs snapshot, send snapshot directly
+    fn prepare_append_delegate(&mut self, to: u64, prs: &mut ProgressSet) -> GroupDelegate {
+        let mut delegate = GroupDelegate::default();
+        if let Some(group) = self.groups.get_members(to) {
+            let mut m = Message::default();
+            m.set_msg_type(MessageType::MsgBroadcast);
+            m.set_from(self.id);
+            // Pick the delegate
+            let group_id = self.groups.get_group_id(to).unwrap();
+            let delegate_id = match self.groups.get_delegate(group_id) {
+                Some(pre_delegate) => {
+                    let delegate_pr = prs.get(pre_delegate).unwrap();
+                    if delegate_pr.is_paused() {
+                        self.pick_delegate(&group, &prs).0
+                    } else {
+                        pre_delegate
+                    }
+                }
+                None => self.pick_delegate(&group, &prs).0,
+            };
+            dbg!(prs.get(4).unwrap());
+            dbg!(delegate_id);
+            dbg!(to);
+            if delegate_id != INVALID_ID && delegate_id != to {
+                self.groups.set_delegate(delegate_id);
+                let target_pr = prs.get_mut(to).unwrap();
+                if target_pr.is_paused() {
+                    trace!(
+                        self.logger,
+                        "Skipping sending to {to}, it's paused",
+                        to = to;
+                        "progress" => ?target_pr,
+                    );
+                    return delegate;
+                } else {
+                    let term = self.raft_log.term(target_pr.next_idx - 1);
+                    let ents = self.raft_log.entries(target_pr.next_idx, self.max_msg_size);
+                    if target_pr.pending_request_snapshot != INVALID_INDEX
+                        || term.is_err()
+                        || ents.is_err()
+                    {
+                        // Let the leader do its job if the target follower needs snapshot
+                        return delegate;
```

Comment:
In the current implementation leader sends snapshots to followers directly, the delegate is bypassed.
However I think we can let the delegate sends snapshots to followers. Here the leader needs only call `pr.become_snapshot`.
And, if the delegate meets `SnapshotTemporarilyUnavailable` when sending the snapshot, it will report it to the leader in `BroadcastResponse`.
After that, leader can call `pr.snapshot_failure()` and `pr.become_probe()` and then probe the follower.